program/rust/src/tests/test_ema.rs

```diff
@@ -1,23 +1,191 @@
+    },
     test_generator::test_resources,
 };
 
+#[test]
+fn test_ema_multiple_publishers_same_slot() -> Result<(), Box<dyn std::error::Error>> {
+    let mut instruction_data = [0u8; size_of::<UpdPriceArgs>()];
+
+    let program_id = Pubkey::new_unique();
+
+    let mut funding_setup = AccountSetup::new_funding();
+    let funding_account = funding_setup.as_account_info();
+
+    let mut price_setup = AccountSetup::new::<PriceAccount>(&program_id);
+    let mut price_account = price_setup.as_account_info();
+    price_account.is_signer = false;
+    PriceAccount::initialize(&price_account, PC_VERSION).unwrap();
+
+    add_publisher(&mut price_account, funding_account.key, 0);
+
+    let mut clock_setup = AccountSetup::new_clock();
+    let mut clock_account = clock_setup.as_account_info();
+    clock_account.is_signer = false;
+    clock_account.is_writable = false;
+
+    update_clock_slot(&mut clock_account, 1);
+
+    populate_instruction(&mut instruction_data, 10, 1, 1);
+    process_instruction(
+        &program_id,
+        &[
+            funding_account.clone(),
+            price_account.clone(),
+            clock_account.clone(),
+        ],
+        &instruction_data,
+    )?;
+
+    {
+        let price_data = load_checked::<PriceAccount>(&price_account, PC_VERSION).unwrap();
+        assert_eq!(price_data.prev_twap_.val_, 0);
+        assert_eq!(price_data.prev_twac_.val_, 0);
+        assert_eq!(price_data.twap_.val_, 10);
+        assert_eq!(price_data.twac_.val_, 1);
+    }
+
+    // add new test for multiple publishers and ensure that ema is updated correctly
+    let mut funding_setup_two = AccountSetup::new_funding();
+    let funding_account_two = funding_setup_two.as_account_info();
+
+    add_publisher(&mut price_account, funding_account_two.key, 1);
+
+    update_clock_slot(&mut clock_account, 2);
+
+    populate_instruction(&mut instruction_data, 20, 1, 2);
+    process_instruction(
+        &program_id,
+        &[
+            funding_account.clone(),
+            price_account.clone(),
+            clock_account.clone(),
+        ],
+        &instruction_data,
+    )?;
+
+    {
+        let price_data = load_checked::<PriceAccount>(&price_account, PC_VERSION).unwrap();
+        assert_eq!(price_data.prev_twap_.val_, 10);
+        assert_eq!(price_data.prev_twac_.val_, 1);
+        assert_eq!(price_data.twap_.val_, 15);
+        assert_eq!(price_data.twac_.val_, 1);
+    }
+
+    populate_instruction(&mut instruction_data, 30, 1, 2);
+    process_instruction(
+        &program_id,
+        &[
+            funding_account_two.clone(),
+            price_account.clone(),
+            clock_account.clone(),
+        ],
+        &instruction_data,
+    )?;
+
+    {
+        let price_data = load_checked::<PriceAccount>(&price_account, PC_VERSION).unwrap();
+        assert_eq!(price_data.prev_twap_.val_, 10);
+        assert_eq!(price_data.prev_twac_.val_, 1);
+        assert_eq!(price_data.twap_.val_, 12);
```

Comment:
why is this value 12? I get why the one above is 15 (10 on the first slot, 20 on the second slot). But then adding a price of 30 seems like it should make this value go up, right? 

I.e., how does the EMA decrease when the aggregate price in the slot increases?

Reply:
the reason for a lower number is the increased confidence. it seems that a new price has the weight of 1/confidence when being added. The formula seems complex and i’m not sure i get it. But if you just add a single price by a publisher with agg 25 and conf 5 then the result will be the same.

Part of the forumla that has it:
```
numer = numer * decay + val * 1/conf
denom = denum * decay + 1/conf
=> val = numer / denom 
```

I think it is worth documenting the reason.